beaker/precompile.py

```diff
@@ -0,0 +1,162 @@
+from dataclasses import dataclass, field
+from pyteal import *
+from algosdk.source_map import SourceMap
+from algosdk.future.transaction import LogicSigAccount
+from algosdk.atomic_transaction_composer import LogicSigTransactionSigner
+from beaker.consts import PROGRAM_DOMAIN_SEPARATOR
+from beaker.logic_signature import LogicSignature
+from beaker.lib.strings import encode_uvarint
+
+
+@dataclass
+class TemplateValue:
+    name: str = field(kw_only=True)
+    is_bytes: bool = field(kw_only=True)
+    line: int = field(kw_only=True)
+    pc: int = 0
+
+
+TMPL_PREFIX = "TMPL_"
+PUSH_BYTES = "pushbytes"
+PUSH_INT = "pushint"
+
+
+def py_encode_uvarint(integer: int) -> bytes:
+    """Encodes an integer as an uvarint.
+    :param integer: the integer to encode
+    :return: bytes containing the integer encoded as an uvarint
+    """
+
+    def to_byte(integer: int) -> int:
+        return integer & 0b1111_1111
+
+    buffer: bytearray = bytearray()
+
+    while integer >= 0b1000_0000:
+        buffer.append(to_byte(integer) | 0b1000_0000)
+        integer >>= 7
+
+    buffer.append(to_byte(integer))
+
+    return bytes(buffer)
+
+
+class Precompile:
+    def __init__(self, lsig: LogicSignature):
+        self.lsig = lsig
+
+        self.program = lsig.program
+
+        self.binary = None
+        self.addr = None
+        self.map = None
+
+        self.template_values: list[TemplateValue] = []
+
+        # Replace the program text with 0 value
```

Comment:
Why with 0? We should replace it with the arg value, shouldn't we?

Reply:
we replace with 0 prior to assembly, later the actual value is patched into the binary directly after encoding it properly